Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp

```diff
@@ -362,56 +382,70 @@ void DiscusMultipleScatteringCorrection::prepareStructureFactors() {
       g_log.information() << "Creating isotropic structure factor for " << matName << std::endl;
-      m_SQWSs.push_back(ComponentWorkspaceMapping{m_env->getComponentPtr(i), matName, isotropicSQ});
+      AddWorkspaceToDiscus2DData(m_env->getComponentPtr(i), matName, isotropicSQ);
     }
   }
+}
 
+/**
+ * Function to convert between a Matrix workspace and the internal simplified 2D data structure. This decouples the
+ * internal calculation logic from the Mantid workspaces
+ */
+void DiscusMultipleScatteringCorrection::AddWorkspaceToDiscus2DData(const Geometry::IObject_const_sptr &shape,
+                                                                    const std::string_view &matName,
+                                                                    API::MatrixWorkspace_sptr SQWS) {
   // avoid repeated conversion of bin edges to points inside loop by converting to point data
-  for (auto &SQWSMapping : m_SQWSs) {
-    auto &SQWS = SQWSMapping.SQ;
-    convertWsBothAxesToPoints(SQWS);
-    // if S(Q,w) has been supplied ensure Q is along the x axis of each spectrum (so same as S(Q))
-    if (SQWS->getAxis(1)->unit()->unitID() == "MomentumTransfer") {
-      auto transposeAlgorithm = this->createChildAlgorithm("Transpose");
-      transposeAlgorithm->initialize();
-      transposeAlgorithm->setProperty("InputWorkspace", SQWS);
-      transposeAlgorithm->setProperty("OutputWorkspace", "_");
-      transposeAlgorithm->execute();
-      SQWS = transposeAlgorithm->getProperty("OutputWorkspace");
-    } else if (SQWS->getAxis(1)->isSpectra()) {
-      // for elastic set w=0 on the spectrum axis to align code with inelastic
-      auto newAxis = std::make_unique<NumericAxis>(std::vector<double>{0.});
-      newAxis->setUnit("DeltaE");
-      SQWS->replaceAxis(1, std::move(newAxis));
-    }
-    SQWSMapping.logSQ = SQWSMapping.SQ->clone();
-    convertToLogWorkspace(SQWSMapping.logSQ);
-  }
+  convertWsBothAxesToPoints(SQWS);
+  // if S(Q,w) has been supplied ensure Q is along the x axis of each spectrum (so same as S(Q))
+  if (SQWS->getAxis(1)->unit()->unitID() == "MomentumTransfer") {
+    auto transposeAlgorithm = this->createChildAlgorithm("Transpose");
+    transposeAlgorithm->initialize();
+    transposeAlgorithm->setProperty("InputWorkspace", SQWS);
+    transposeAlgorithm->setProperty("OutputWorkspace", "_");
+    transposeAlgorithm->execute();
+    SQWS = transposeAlgorithm->getProperty("OutputWorkspace");
+  } else if (SQWS->getAxis(1)->isSpectra()) {
+    // for elastic set w=0 on the spectrum axis to align code with inelastic
+    auto newAxis = std::make_unique<NumericAxis>(std::vector<double>{0.});
+    newAxis->setUnit("DeltaE");
+    SQWS->replaceAxis(1, std::move(newAxis));
+  }
+  auto specAxis = dynamic_cast<NumericAxis *>(SQWS->getAxis(1));
+  std::vector<DiscusData1D> data;
+  for (size_t i = 0; i < SQWS->getNumberHistograms(); i++) {
+    data.push_back(DiscusData1D{SQWS->histogram(i).dataX(), SQWS->histogram(i).dataY()});
```

Comment:
Would it be better to use `emplace_back` here and just provide the arguments of the `DiscusData1D` constructor? I think that would then avoid a move

Reply:
Good spot. I'll change to emplace_back. The compiler seems to want a constructor added to DiscusData1D that takes the 2 parameters but that's no big deal to add. I seem to have used push_back elsewhere in this unit as well. The only other instance that looks wrong is in DiscusMultipleScatteringCorrection::updateWeightAndPosition so I'll change that to emplace_back also

Reply:
Great, thank you!